### PR TITLE
[SPARK-26736][SQL] Partition pruning through nondeterministic expressions in Hive tables

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR intends to improve partition pruning for nondeterministic expressions in Hive tables:

Before this PR:
```
scala> sql("""create table test(id int) partitioned by (dt string)""")
scala> sql("""select * from test where dt='20190101' and rand() < 0.5""").explain()

== Physical Plan ==
*(1) Filter ((isnotnull(dt#19) AND (dt#19 = 20190101)) AND (rand(6515336563966543616) < 0.5))
+- Scan hive default.test [id#18, dt#19], HiveTableRelation `default`.`test`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [id#18], [dt#19], Statistics(sizeInBytes=8.0 EiB)
```
After this PR:
```
== Physical Plan ==
*(1) Filter (rand(-9163956883277176328) < 0.5)
+- Scan hive default.test [id#0, dt#1], HiveTableRelation `default`.`test`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [id#0], [dt#1], Statistics(sizeInBytes=8.0 EiB), [isnotnull(dt#1), (dt#1 = 20190101)]
```
This PR is the rework of #24118.

### Why are the changes needed?

For better performance.

### Does this PR introduce any user-facing change?

No.

### How was this patch tested?

Unit tests added.

### CODE REVIEW

**Comment by maropu** [comment]
cc: @cloud-fan @Ngone51

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala [hunk 1]
```diff
@@ -141,6 +141,21 @@ class PruningSuite extends HiveComparisonTest with BeforeAndAfter {
...
+  createPruningTest("Partition pruning - with filter containing non-deterministic condition",
+    "SELECT value, hr FROM srcpart1 WHERE ds = '2008-04-08' AND hr < 12 AND rand() < 1",
+    Seq("value", "hr"),
+    Seq("value", "hr"),
+    Seq(
+      Seq("2008-04-08", "11")))
+
+  createPruningTest(
+    "Partition pruning - with filter containing non-deterministic condition in sub And-expr",
+    "SELECT value, hr FROM srcpart1 WHERE (ds = '2008-04-08' AND rand() < 1) AND (hr < 12)",
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala]: not sure we need this. As a UT, we know that it doesn't matter how predicates are ANDed together.

Comment (maropu) [reply sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/PruningSuite.scala]: ok, I will drop.

**Review by cloud-fan — APPROVED** [review]

**Comment by maropu** [comment]
Merged to master.

**Comment by dongjoon-hyun** [comment]
+1, late LGTM.

**Pull request closed without merging**